stratosphere/creport/source/creport_crash_report.cpp

```diff
@@ -14,107 +14,171 @@
 
-    if (this->code_list.code_count) {
-        ctx->aarch64_ctx.start_address = this->code_list.code_infos[0].start_address;
+            /* Process dying message for applications. */
+            if (this->IsApplication()) {
+                this->ProcessDyingMessage();
+            }
+
+            /* Nintendo's creport finds extra modules by looking at all threads if application, */
+            /* but there's no reason for us not to always go looking. */
+            for (size_t i = 0; i < this->thread_list.GetThreadCount(); i++) {
+                this->module_list.FindModulesFromThreadInfo(this->debug_handle, this->thread_list.GetThreadInfo(i));
+            }
+
+            /* Nintendo's creport saves the report to erpt here, but we'll save to SD card later. */
+        }
     }
 
-    /* For ams fatal... */
-    ctx->aarch64_ctx.afsr0 = this->process_info.title_id;
+    void CrashReport::GetFatalContext(FatalContext *out) const {
+        std::memset(out, 0, sizeof(*out));
 
-    return ctx;
-}
+        /* TODO: Support generating 32-bit fatal contexts? */
+        out->is_aarch32 = false;
+        out->type = static_cast<u32>(this->exception_info.type);
+
+        for (size_t i = 0; i < 29; i++) {
+            out->aarch64_ctx.x[i] = this->crashed_thread.GetGeneralPurposeRegister(i);
+        }
+        out->aarch64_ctx.fp = this->crashed_thread.GetFP();
+        out->aarch64_ctx.lr = this->crashed_thread.GetLR();
+        out->aarch64_ctx.pc = this->crashed_thread.GetPC();
+
+        out->aarch64_ctx.stack_trace_size = this->crashed_thread.GetStackTraceSize();
+        for (size_t i = 0; i < out->aarch64_ctx.stack_trace_size; i++) {
+            out->aarch64_ctx.stack_trace[i] = this->crashed_thread.GetStackTrace(i);
+        }
 
-void CrashReport::ProcessExceptions() {
-    if (!IsOpen()) {
-        return;
+        if (this->module_list.GetModuleCount()) {
+            out->aarch64_ctx.start_address = this->module_list.GetModuleStartAddress(0);
+        }
+
+        /* For ams fatal, which doesn't use afsr0, pass title_id instead. */
+        out->aarch64_ctx.afsr0 = this->process_info.title_id;
     }
 
-    DebugEventInfo d;
-    while (R_SUCCEEDED(svcGetDebugEvent((u8 *)&d, this->debug_handle))) {
-        switch (d.type) {
-            case DebugEventType::AttachProcess:
-                HandleAttachProcess(d);
-                break;
-            case DebugEventType::Exception:
-                HandleException(d);
-                break;
-            case DebugEventType::AttachThread:
-                HandleAttachThread(d);
-            case DebugEventType::ExitProcess:
-            case DebugEventType::ExitThread:
-            default:
-                break;
+    void CrashReport::ProcessExceptions() {
+        /* Loop all debug events. */
+        svc::DebugEventInfo d;
+        while (R_SUCCEEDED(svcGetDebugEvent(reinterpret_cast<u8 *>(&d), this->debug_handle))) {
```

Comment:
This gets me thinking. Should libnx svcGetDebugEvent accept void* so that you don't need to do this ugly cast?

Reply:
that would be pretty nice.